### PR TITLE
Start of USB host API

### PR DESCRIPTION
**Pull request by tannewt**
This allows you to list and explore connected USB devices. It
only stubs out the methods to communicate to endpoints. That will
come in a follow up once TinyUSB has it. (It's in progress.)

Related to #5986

### CODE REVIEW

**Comment by tannewt** [comment]
The last build was mostly network failures and one size failure that seems to be unrelated. Please review and I'll likely need to do some tweaks that will lead to another CI run.

File: shared-module/usb/core/Device.c [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+            uint32_t hc = ((chr & 0xFFFF0000) - 0xD8000000) >> 6;    /* Get high 10 bits */
+            chr = (chr & 0xFFFF) - 0xDC00;                  /* Get low 10 bits */
+            chr = (hc | chr) + 0x10000;
+            *utf8++ = (uint8_t)(0xF0 | (chr >> 18 & 0x07));
+            *utf8++ = (uint8_t)(0x80 | (chr >> 12 & 0x3F));
+            *utf8++ = (uint8_t)(0x80 | (chr >> 6 & 0x3F));
+            *utf8++ = (uint8_t)(0x80 | (chr >> 0 & 0x3F));
+        }
+    }
+}
```
Comment (dhalbert) [inline shared-module/usb/core/Device.c:99]: (Eventually?) factor this out into some utf16 util file.

File: shared-module/usb/core/Device.c [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        } else if (chr < 0x800) {
+            total_bytes += 2;
+        } else if (chr < 0x10000) {
+            total_bytes += 3;
+        } else {
+            total_bytes += 4;
+        }
+    }
+    return total_bytes;
+}
```
Comment (dhalbert) [inline shared-module/usb/core/Device.c:117]: Factor into utf16 file.

File: shared-module/usb/core/Device.c [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    }
+    size_t size = _count_utf8_bytes(temp_buf + 1, utf16_len);
+    vstr_t vstr;
+    vstr_init_len(&vstr, size + 1);
+    byte *p = (byte *)vstr.buf;
+    // Null terminate.
+    p[size] = '\0';
+    _convert_utf16le_to_utf8(temp_buf + 1, utf16_len, p, size);
+    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+}
```
Comment (dhalbert) [inline shared-module/usb/core/Device.c:141]: factor into utf16 file

File: shared-bindings/usb/core/Device.c [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+//|
+STATIC mp_obj_t usb_core_device_ctrl_transfer(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_bmRequestType, ARG_bRequest, ARG_wValue, ARG_wIndex, ARG_data_or_wLength, ARG_timeout };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_bmRequestType, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_bRequest, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_wValue, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_wIndex, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_data_or_wLength, MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_timeout, MP_ARG_INT, {.u_int = 0} },
```
[the remark this answers is not in the document]
Comment (tannewt) [reply shared-bindings/usb/core/Device.c:238]: What do you mean? I assume the converter code would raise an exception when given None.

File: shared-module/usb/core/Device.c [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    return pid;
+}
+
+STATIC xfer_result_t _get_string_result;
+STATIC bool _transfer_done_cb(uint8_t daddr, tusb_control_request_t const *request, xfer_result_t result) {
+    (void)daddr;
+    (void)request;
+    _get_string_result = result;
+    return true;
+}
```
[the remark this answers is not in the document]
Comment (tannewt) [reply shared-module/usb/core/Device.c:69]: Added a comment. It shouldn't need any other logic.

[the remark this answers is not in the document]
Comment (dhalbert) [reply shared-module/usb/core/Device.c:69]: I wasn't sure if it was just the callback logic or not.

File: shared-bindings/usb/core/Device.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhalbert) [reply shared-bindings/usb/core/Device.c:238]: The signature says the default is `None`, but if you tried to pass an explicit `timeout=None`, you'd get an error, despite what the signature says.

**Comment by dhalbert** [comment]
One board is not fitting.

**Comment by tannewt** [comment]
TinyUSB grew in size:

```
_usbd_rhport                                                      +1
dcd_mutex_def                                                     +2
dcd_mutex                                                         +4
usbd_int_set                                                     +20
storage_erase_filesystem                                          +8
complex_attr                                                      +2
osal_queue_send.isra.0                                           -16
edpt_dma_start                                                   +36
compile_decorated                                                 -4
mp_obj_exception_attr                                             +4
usbd_edpt_xfer.isra.0                                            +80
compile_atom_expr_normal                                          +4
tusb_hal_nrf_power_event                                          +4
mscd_xfer_cb                                                     +24
mp_qstr_const_attr                                                -4
usb_background                                                   -16
main                                                             +24
mp_qstr_const_pool                                                -8
_usbd_initialized                                                 -1

Total 258461 - 258297                                           +164
```

@jpconstantineu this disables the traceback module on the bluemicro833. It's new so I figured no code uses it. (Could do watchdog as an alternative.)

**Review by dhalbert — APPROVED** [review]
:+1:

**Pull request merged**